smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,308 @@
+
+            // If all endpoints have been tried and connection couldn't be established, failure would be returned, null otherwise.
+            StateTransitionResult.Failure failure = connectionAttemptState.establishWebsocketConnection();
+            if (failure != null) {
+                return failure;
+            }
+
+            connectionInternal.setTransport(websocketTransport);
+
+            WebsocketRemoteConnectionEndpoint connectedEndpoint = connectionAttemptState.getConnectedEndpoint();
+
+            // Construct a WebsocketConnectedResult using the connected endpoint.
+            return new WebsocketConnectedResult(connectedEndpoint);
+        }
+    }
+
+    public EstablishingWebsocketConnectionState constructEstablishingWebsocketConnectionState(
+                    EstablishingWebsocketConnectionStateDescriptor establishingWebsocketConnectionStateDescriptor,
+                    ModularXmppClientToServerConnectionInternal connectionInternal) {
+        return new EstablishingWebsocketConnectionState(establishingWebsocketConnectionStateDescriptor,
+                        connectionInternal);
+    }
+
+    public static final class WebsocketConnectedResult extends StateTransitionResult.Success {
+        final WebsocketRemoteConnectionEndpoint connectedEndpoint;
+
+        public WebsocketConnectedResult(WebsocketRemoteConnectionEndpoint connectedEndpoint) {
+            super("Websocket connection establised with endpoint: " + connectedEndpoint.getWebsocketEndpoint());
+            this.connectedEndpoint = connectedEndpoint;
+        }
+    }
+
+    private DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints;
+
+    /**
+     * Transport class for {@link ModularXmppClientToServerConnectionModule}'s websocket implementation.
+     */
+    public final class XmppWebsocketTransport extends XmppClientToServerTransport {
+
+        protected XmppWebsocketTransport(ModularXmppClientToServerConnectionInternal connectionInternal) {
+            super(connectionInternal);
+        }
+
+        @Override
+        protected void resetDiscoveredConnectionEndpoints() {
+            discoveredWebsocketEndpoints = null;
+        }
+
+        @Override
+        protected List<SmackFuture<LookupConnectionEndpointsResult, Exception>> lookupConnectionEndpoints() {
+            // Assert that there are no stale discovered endpoints prior performing the lookup.
+            assert discoveredWebsocketEndpoints == null;
+
+            List<SmackFuture<LookupConnectionEndpointsResult, Exception>> futures = new ArrayList<>(2);
+
+            InternalSmackFuture<LookupConnectionEndpointsResult, Exception> websocketEndpointsLookupFuture = new InternalSmackFuture<>();
+
+            connectionInternal.asyncGo(() -> {
+
+                WebsocketRemoteConnectionEndpoint providedEndpoint = null;
+
+                // Check if there is a websocket endpoint already configured.
+                if (moduleDescriptor.getUri() != null) {
```

Comment:
Do not call getUri() twice! Call it once and safe the result in a local variable.